src/1-theory/3-input-and-output.adoc

```diff
@@ -0,0 +1,593 @@
+doing a "Well Actually..." and should spend their time working on something more
+useful. If you are not using HTTP to its full extent, that stance might be
+partially true, but when you start to leverage the full power of HTTP that
+opinion is totally false.
+
+For example, many tools will know that they can cache a GET request, but know
+that caching a POST would be spooky and problematic. They will know they can
+retry a PUT or a DELETE, but better not retry a PATCH. These conventions allow a
+lot of tools that do not have direct knowledge of each other to work together.
+
+Let's learn a bit more about these HTTP methods.
+
+==== GET
+
+Just fetch some stuff from the server. This is the best example of an "idempotent" action, and this is a word which pops up a bit in API land.
+
+Idempotent means you can do this thing over and over again, and it happening two
+or more times won't result in  different outcomes. If you GET the thing, but the
+request fails or times out, and you GET it again, the end result is that you got
+it. Nothing was deleted, or removed, or changed in any lasting way, so if this
+thing is got a bunch of times it is the same as being got once.
+
+NOTE: They can technically have a body some servers and tools will start
+acting a bit wonky, meaning people generally avoid creating GETs with bodies.
+
+==== POST
+
+Often used purely for "creates" but can be used for any non-idempotent action.
+
+Creating something is a good example of a non-idempotent action, because if you
+send two requests to create a thing, you now have two things. You could also be
+triggering the sending of an email, paying an invoice, etc., it does not have to
+be a create.
+
+==== PUT
+
+Often incorrectly associated with being an "edit" action, PUT can actually do a
+whole lot more than that. PUT is designed to be an idempotent way to send data
+to a server, where the request contains the entirety of the resource. Whatever
+it says should go, so if the attempt fails for any reason and the client
+retries, there wil be no negative consequence as its said the same thing twice.
+
+An example of this would be an image upload. An API might have the ability to
+upload an image for a user, which is probably a profile image. A request with
+`PUT /users/jane/image` and a body of either the image contents (or a JSON
+payload with a URL) could then provide the image. It does not matter if the user
+already had an image or not, if the request is a success they will have one. If
+the upload fails that is fine, another request can be set, and it will be
+overridden.
+
+NOTE: Some folks get a bit concerned about this being a "create or update"
+action, but their concern comes from a misplaced sense that HTTP verbs
+correspond to a specific CRUD action. If somebody complains at you about this,
+politely point them at this section.
+
+PUT is not always appropriate, and can lead to race conditions if not used
```

Comment:
Would be cool to include an example of a race condition resulting from PUT (or just explain a bit more) :)

Reply:
^ ignore that,  I should clearly read the section ahead before commenting 🤣 🙈